sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,458 @@
+function, and begins the first phase of the barrier.
+
+[source,c++]
+----
+[[nodiscard]] arrival_token arrive(ptrdiff_t update = 1);
+----
+_Effects_: The calling thread of execution arrives at the barrier and decreases
+the expected count by `update`.
+
+_Returns_: An `arrival_token` that can be passed to `wait`.
+
+[source,c++]
+----
+void wait(arrival_token&& arrival) const;
+----
+_Effects_: Blocks the calling thread of execution until the end of the barrier
+phase associated with `arrival`.
+
+[NOTE]
+====
+Because calling `wait` blocks the calling thread of execution, it may prevent
+forward progress. Unlike `group_barrier`, the member functions of `barrier` do
+not provide additional scheduling guarantees; it is a user's responsibility to
+ensure that calls to `wait` are compatible with the forward progress guarantees
+provided by an implementation.
+====
+
+[source,c++]
+----
+void arrive_and_wait();
+----
+_Effects_: Equivalent to `wait(arrive())`.
+
+[source,c++]
+----
+void arrive_and_drop();
+----
+_Effects_: The calling thread of execution arrives at the barrier, decreases
+the number of threads of execution expected in the next phase, and decreases
+the expected count of the current phase by 1.
+
+
+==== Usage examples
+
+As noted above, `barrier` objects must be allocated in memory that is visible
+to all of the work-items using the barrier. There are many ways to satisfy this
+condition. The examples in this section demonstrate some common ways to
+allocate and construct barrier objects for different scopes.
+
+
+===== Work-group scope
+
+Barriers at work-group scope can be allocated in group-local memory. The
+simplest way to allocate and construct a `barrier` object in group-local memory
+is to use the
+link:../supported/sycl_ext_oneapi_local_memory.asciidoc[sycl_ext_oneapi_local_memory]
+extension.
+
+[source,c++]
+----
+using work_group_barrier = syclex::barrier<sycl::memory_scope::work_group>;
+
+q.parallel_for(..., [=](sycl::nd_item it) {
+
+  // Allocate memory for and construct the barrier
+  auto* bar = sycl::ext::oneapi::group_local_memory<work_group_barrier>(it.get_group(), nthreads);
+
+  // Use the barrier
+  bar->arrive_and_wait();
+
+  // The barrier is automatically destructed and the memory is freed upon kernel completion
+
+}).wait();
+----
+
+
+===== Device scope
+
+Barriers at device scope must be allocated in global memory (including various
+kinds of USM). It is recommended that device scope barriers are initialized on
+the device that will use the barrier.
+
+[source,c++]
+----
+using device_barrier = syclex::barrier<sycl::memory_scope::device>;
+
+// Allocate memory for the barrier
+device_barrier* bar = sycl::device_malloc<device_barrier>(1, q);
+
+// Construct the barrier using placement new
+q.single_task([=]() {
+  new (bar) device_barrier(nthreads);
+}).wait();
+
+// Use the barrier
+q.parallel_for(..., [=](sycl::nd_item it) {
+  auto mybar = std::launder(bar);
```

Comment:
My understanding here is pretty shaky, and after talking to @rolandschulz we weren't 100% sure if `std::launder` is actually required here.  But our conclusion was that a strict reading of the C++ specification says that it _is_ required -- we could avoid this if SYCL said something like "all USM pointers are assumed to be laundered upon launching a kernel", but we don't know if that would have any unintended consequences!  That being the case, I'd prefer to leave the examples as-is for now... but we should probably open an issue somewhere to explore this further.

My understanding below.

In the device barrier case, the barrier isn't being accessed via the pointer that was returned from `new`.  That means we have to use `std::launder` to guarantee that `mybar` is a valid pointer to a barrier object.  We can't use `bar` directly (even though both pointers will, in practice, have the same address).

In the system barrier case, we can overwrite the `bar` pointer with the result of `new` before we enter any of the kernels.  The pointer that gets captured by the kernels is then guaranteed to be a valid pointer to a barrier object.

I know this looks horrible, but it's actually much better than the alternatives we considered.  Because kernel lambdas can't return values, getting the pointer returned by `new` back to the host requires going through an extra USM allocation (i.e. a pointer to a `device_barrier*`).  I did sketch it out, but I found it much harder to follow than the `std::launder` example.

Reply:
I assume that the same question also applies to virtual functions? I imagine that the main method to create a polymorphic object would be through placement new with device or shared allocation passed to it. Meaning that kernel which accesses that object does not access it through a pointer which was returned through `new`

Reply:
I think you may be right, but tagging @rolandschulz to comment.